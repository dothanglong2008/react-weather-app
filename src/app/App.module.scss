* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

html {
  font-size: calc(10px + 1vmin);
}

ul {
  list-style: none;
}

img,
picture,
figure {
  width: 100%;
  object-fit: none;
}

button,
a {
  cursor: pointer;
}

.app {
  background-color: #ebebeb;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}
