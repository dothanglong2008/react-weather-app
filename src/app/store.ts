import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import weatherReducer from "../features/weather/weatherSlice";
import searchFormReducer from "../components/weather/search/searchFormSlice";

export const store = configureStore({
  reducer: {
    weather: weatherReducer,
    searchForm: searchFormReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppStore = typeof store;

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
