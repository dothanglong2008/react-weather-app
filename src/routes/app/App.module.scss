$background-color-primary: #ffffff;
$text-color-primary: #000000;
$font-size-primary: 16px;

* {
  margin: 0;
  padding: 0;
  width: 100%;
}

body {
  background-color: $background-color-primary;
  color: $text-color-primary;
  font-size: $font-size-primary;
}

ul {
  list-style: none;
}

img,
picture,
video {
  max-width: 100%;
}
