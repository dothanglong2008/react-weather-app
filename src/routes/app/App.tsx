import Item from "../../components/weather/single-item/item/Item";
import styles from "./App.module.scss";

function App() {
  return (
    <div className={styles.app}>
      <Item />
    </div>
  );
}

export default App;
