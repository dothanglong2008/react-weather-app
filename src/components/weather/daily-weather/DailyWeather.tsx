import { useDeferredValue, useMemo, useState } from "react";
import { useAppSelector } from "../../../app/hooks";
import useFetchWeatherByDays from "../../../hooks/useFetchWeatherByDays";
import { selectWeatherData } from "../../../features/weather/weatherSlice";
import SectionHeading from "../../ui/elements/section-heading/SectionHeading";
import styles from "./DailyWeather.module.scss";
import Item from "../item/Item";
import { DAYS_GAP, MAX_DAYS, MIN_DAYS } from "../../../constants/constants";
import TextBox from "../../ui/elements/text-box/TextBox";

function DailyWeather() {
  const [days, setDays] = useState(5);
  const weatherData = useAppSelector(selectWeatherData);
  useFetchWeatherByDays(days);
  const deferWeatherData = useDeferredValue(weatherData);

  const showMore = () => {
    setDays((days) => days + DAYS_GAP);
  };

  const showLess = () => {
    setDays((days) => days - DAYS_GAP);
  };

  let dailyForecast = useMemo(() => {
    if (deferWeatherData.current.last_updated_epoch === 0) {
      return <TextBox content={<>Loading...</>} />;
    } else if (
      deferWeatherData.status === "succeeded" ||
      deferWeatherData.status === "loading"
    ) {
      return (
        <>
          <ul>
            {deferWeatherData?.forecast?.forecastday.map((forecastDay) => {
              const table = [
                { field: "Min Temperature", value: forecastDay.day.maxtemp_c },
                { field: "Max Temperature", value: forecastDay.day.maxtemp_c },
                { field: "Max Wind", value: forecastDay.day.maxwind_kph },
                { field: "Humidity", value: forecastDay.day.avghumidity },
                {
                  field: "Chance Of Rain",
                  value: forecastDay.day.daily_chance_of_rain,
                },
                {
                  field: "Precipitation",
                  value: forecastDay.day.totalprecip_mm,
                },
                { field: "UV", value: forecastDay.day.uv },
              ];
              return (
                <Item
                  key={forecastDay.date_epoch}
                  time={forecastDay.date}
                  icon={forecastDay.day.condition.icon}
                  temp={forecastDay.day.avgtemp_c}
                  humidity={forecastDay.day.avghumidity}
                  chanceOfRain={forecastDay.day.daily_chance_of_rain}
                  extraTable={table}
                />
              );
            })}
          </ul>
          {days > MIN_DAYS && (
            <button className={styles.show} onClick={showLess}>
              Show less
            </button>
          )}
          {days < MAX_DAYS && (
            <button className={styles.show} onClick={showMore}>
              Show more
            </button>
          )}
        </>
      );
    } else if (deferWeatherData.status === "failed") {
      return <p>No data to display</p>;
    }
  }, [deferWeatherData, days]);

  return (
    <>
      <SectionHeading
        content={
          <>
            Next {days} Days Weather Forecast For{" "}
            {deferWeatherData.location.name}
          </>
        }
      />
      {dailyForecast}
    </>
  );
}

export default DailyWeather;
