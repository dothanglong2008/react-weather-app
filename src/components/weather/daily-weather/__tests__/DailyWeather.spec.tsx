import renderer from "react-test-renderer";
import DailyWeather from "../DailyWeather";
import { Provider } from "react-redux";
import { store } from "../../../../app/store";

describe("Test DailyWeather component", () => {
  it("renders DailyWeather correctly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <DailyWeather />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
