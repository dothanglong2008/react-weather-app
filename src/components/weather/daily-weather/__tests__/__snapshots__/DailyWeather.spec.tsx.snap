// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DailyWeather component renders DailyWeather correctly 1`] = `
Array [
  <h2
    className="sectionHeading"
  >
    Next 
    5
     Days Weather Forecast For
     
  </h2>,
  <p>
    Loading...
  </p>,
]
`;
